optuna/samplers/_nsga2/crossover.py

```diff
@@ -0,0 +1,431 @@
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import List
+from typing import Sequence
+
+import numpy as np
+
+from optuna._transform import _SearchSpaceTransform
+from optuna.distributions import BaseDistribution
+from optuna.distributions import CategoricalDistribution
+from optuna.distributions import DiscreteUniformDistribution
+from optuna.distributions import IntLogUniformDistribution
+from optuna.distributions import IntUniformDistribution
+from optuna.distributions import LogUniformDistribution
+from optuna.distributions import UniformDistribution
+from optuna.study import Study
+from optuna.study import StudyDirection
+from optuna.trial import FrozenTrial
+
+
+_NUMERICAL_DISTRIBUTIONS = (
+    UniformDistribution,
+    LogUniformDistribution,
+    DiscreteUniformDistribution,
+    IntUniformDistribution,
+    IntLogUniformDistribution,
+)
+
+
+def crossover(
+    crossover_name: str,
+    study: Study,
+    parent_population: Sequence[FrozenTrial],
+    search_space: Dict[str, BaseDistribution],
+    rng: np.random.RandomState,
+    swapping_prob: float,
+    dominates: Callable[[FrozenTrial, FrozenTrial, Sequence[StudyDirection]], bool],
+) -> Dict[str, Any]:
+
+    numerical_search_space: Dict[str, BaseDistribution] = {}
+    numerical_distributions: List[BaseDistribution] = []
+
+    for key, value in search_space.items():
+        if isinstance(value, _NUMERICAL_DISTRIBUTIONS):
+            numerical_search_space[key] = value
+            numerical_distributions.append(value)
+
+    if len(numerical_distributions) != 0:
+        numerical_transform = _SearchSpaceTransform(numerical_search_space)
+
+    while True:  # Repeat while parameters lie outside search space boundaries.
```

Comment:
IMO, this while loop looks too large to grasp the overview and structure of the block without careful reading.
I think it's okay not to address that in this PR but this loop could be simplified by factoring out each sub-block into another function as follows:
```python
    while True:
        child_params = _categorical_params_crossover(search_space, swapping_prob)  # Replace L53..L64
        if len(numerical_distributions) == 0:
            if _is_contained(child_params, search_space):
                break
            continue

        # The following is applied only for numerical parameters.
        parents_numerical_params_array = _transform_numerical_search_space( # Replace L72..L82
           numerical_transform, parents
        )

         if crossover_name == "uniform":
             child_params_array = _uniform_crossover(  # Replace L84..L89
                 parents_numerical_params_array, rng, swapping_prob
             )
        elif crossover_name == "blxalpha":
            child_params_array = _blxalpha_crossover(  # Replace L91..L94
                parents_numerical_params_array, rng
             )
        elif crossover_name == "sbx":
            child_params_array = _sbx_crossover(  # Replace L96..L107
                parents_numerical_params_array, rng, study, numerical_distributions
             )
        elif crossover_name == "vsbx":
            child_params_array = _vsbx_crossover(  # Replace L108..L114
                parents_numerical_params_array, rng, study
             )
        elif crossover_name == "undx":
            child_params_array = _undx_crossover(  # Replace L116..L125
                parents_numerical_params_array, rng
             )
        elif crossover_name == "spx":
            child_params_array = _undx_crossover(  # Replace L127..128
                parents_numerical_params_array, rng
             )
        else:
            assert False

        child_numerical_params = numerical_transform.untransform(child_params_array)
        child_params.update(child_numerical_params)

        if _is_contained(child_params, search_space):
            break
```

Reply:
The code is pretty straight forward so I personally think that the current code is fine, especially if it's going to be refactored into classes, but I do agree it's long and +1 for the change. If you don't want to introduce indirections for each algorithm, or change multiple places on the other hand, I think an alternative is to simply factor out the body of the loop into its own function. Definitely not strong opinions, mostly dropped by for a quick comment.

Reply:
@hvy Thank you for your comment. "to simply factor out the body of the loop" sounds reasonable.

Reply:
Thank you for your comment.

>> I think an alternative is to simply factor out the body of the loop into its own function. Definitely not strong opinions
Does this mean that each crossover function (e.g. `_uniform`, `_blxalpha`) will have a while loop process(L52 ... L136)?

Reply:
> Does this mean ...

No, I had imagined the code like the following:
```python
# New function.
def try_crossover(...) -> Dict[str, Any]:
    ... # Copy code from L53 ~ L133 (but `break` should be replaced with `return child_params`)
    return child_params

def crossover(...) -> Dict[str, Any]:
    ...
    while True:
        child_params = try_crossover(...)
        if _is_contained(child_params, search_space):
            break
    ...
```

Reply:
Thank you!
I understand now. I've pushed the fixed version.

Reply:
Thanks, looks good! Let me resolve this thread as addressed. 
@sile Please reopen if that'd be relevant.